File: js/swing-levels.js
/* Swing high / low (pivot) level detection */
const SwingLevels = (() => {
  function isPivotHigh(highs, index, pivotBars) {
    if (index < pivotBars || index >= highs.length - pivotBars) return false;
    const level = highs[index];
    for (let j = index - pivotBars; j <= index + pivotBars; j++) {
      if (j !== index && highs[j] >= level) return false;
    }
    return true;
  }

  function isPivotLow(lows, index, pivotBars) {
    if (index < pivotBars || index >= lows.length - pivotBars) return false;
    const level = lows[index];
    for (let j = index - pivotBars; j <= index + pivotBars; j++) {
      if (j !== index && lows[j] <= level) return false;
    }
    return true;
  }

  function recentSwingLevels(highs, lows, endIndex, { pivotBars, lookback }) {
    const start = Math.max(pivotBars, endIndex - lookback);
    const searchEnd = endIndex - pivotBars;
    if (searchEnd < start) {
      return { swingHigh: null, swingLow: null, swingHighIndex: null, swingLowIndex: null };
    }

    let swingHigh = null;
    let swingLow = null;
    let swingHighIndex = null;
    let swingLowIndex = null;

    for (let i = searchEnd; i >= start; i--) {
      if (swingHigh == null && isPivotHigh(highs, i, pivotBars)) {
        swingHigh = highs[i];
        swingHighIndex = i;
      }
      if (swingLow == null && isPivotLow(lows, i, pivotBars)) {
        swingLow = lows[i];
        swingLowIndex = i;
      }
      if (swingHigh != null && swingLow != null) break;
    }

    return { swingHigh, swingLow, swingHighIndex, swingLowIndex };
  }

  function nearLevel(price, level, tolerancePct) {
    if (level == null || level <= 0) return false;
    return (Math.abs(price - level) / level) * 100 <= tolerancePct;
  }

  function aboveLevel(price, level) {
    return level != null && price > level;
  }

  function belowLevel(price, level) {
    return level != null && price < level;
  }

  function swingLongOk(price, swingHigh, swingLow, settings) {
    if (!settings.useSwingLevels) return true;
    const mode = settings.swingMode || 'bounce';
    if (mode === 'breakout') return aboveLevel(price, swingHigh);
    return nearLevel(price, swingLow, settings.swingNearPct);
  }

  function swingShortOk(price, swingHigh, swingLow, settings) {
    if (!settings.useSwingLevels) return true;
    const mode = settings.swingMode || 'bounce';
    if (mode === 'breakout') return belowLevel(price, swingLow);
    return nearLevel(price, swingHigh, settings.swingNearPct);
  }

  function buildPivotMarkers(candles, levels) {
    const { swingHigh, swingLow, swingHighIndex, swingLowIndex } = levels;
    const markers = [];
    if (swingHighIndex != null && swingHigh != null && candles[swingHighIndex]) {
      markers.push({
        time: candles[swingHighIndex].time,
        position: 'aboveBar',
        color: '#ef5350',
        shape: 'circle',
        text: `H ${Math.round(swingHigh)}`,
      });
    }
    if (swingLowIndex != null && swingLow != null && candles[swingLowIndex]) {
      markers.push({
        time: candles[swingLowIndex].time,
        position: 'belowBar',
        color: '#26a69a',
        shape: 'circle',
        text: `L ${Math.round(swingLow)}`,
      });
    }
    return markers;
  }

  function calcFromCandles(candles, settings) {
    if (!candles?.length) {
      return { swingHigh: null, swingLow: null, swingHighIndex: null, swingLowIndex: null };
    }
    const highs = candles.map((c) => c.high);
    const lows = candles.map((c) => c.low);
    return recentSwingLevels(highs, lows, candles.length - 1, {
      pivotBars: settings.swingPivotBars || 5,
      lookback: settings.swingLookback || 50,
    });
  }

  function calcStopPrice(side, swingHigh, swingLow, bufferPct) {
    if (side === 'LONG' && swingLow != null) {
      return Math.round(swingLow * (1 - bufferPct / 100) * 100) / 100;
    }
    if (side === 'SHORT' && swingHigh != null) {
      return Math.round(swingHigh * (1 + bufferPct / 100) * 100) / 100;
    }
    return null;
  }

  function calcStopFromCandles(candles, side, settings) {
    if (!settings.useSwingStopLoss || !candles?.length) return null;
    const levels = calcFromCandles(candles, settings);
    return calcStopPrice(
      side,
      levels.swingHigh,
      levels.swingLow,
      settings.swingStopBufferPct ?? 0.2,
    );
  }

  return {
    isPivotHigh,
    isPivotLow,
    recentSwingLevels,
    nearLevel,
    aboveLevel,
    belowLevel,
    swingLongOk,
    swingShortOk,
    buildPivotMarkers,
    calcFromCandles,
    calcStopPrice,
    calcStopFromCandles,
  };
})();

window.SwingLevels = SwingLevels;

File: bot/swing_levels.py
"""Swing high / low (pivot) level detection."""

from __future__ import annotations


def is_pivot_high(highs: list[float], index: int, pivot_bars: int) -> bool:
    if index < pivot_bars or index >= len(highs) - pivot_bars:
        return False
    level = highs[index]
    for j in range(index - pivot_bars, index + pivot_bars + 1):
        if j != index and highs[j] >= level:
            return False
    return True


def is_pivot_low(lows: list[float], index: int, pivot_bars: int) -> bool:
    if index < pivot_bars or index >= len(lows) - pivot_bars:
        return False
    level = lows[index]
    for j in range(index - pivot_bars, index + pivot_bars + 1):
        if j != index and lows[j] <= level:
            return False
    return True


def recent_swing_levels(
    highs: list[float],
    lows: list[float],
    end_index: int,
    *,
    pivot_bars: int,
    lookback: int,
) -> tuple[float | None, float | None]:
    """Most recent confirmed swing high and low before end_index."""
    start = max(pivot_bars, end_index - lookback)
    search_end = end_index - pivot_bars
    if search_end < start:
        return None, None

    swing_high: float | None = None
    swing_low: float | None = None

    for i in range(search_end, start - 1, -1):
        if swing_high is None and is_pivot_high(highs, i, pivot_bars):
            swing_high = highs[i]
        if swing_low is None and is_pivot_low(lows, i, pivot_bars):
            swing_low = lows[i]
        if swing_high is not None and swing_low is not None:
            break

    return swing_high, swing_low


def near_level(price: float, level: float | None, tolerance_pct: float) -> bool:
    if level is None or level <= 0:
        return False
    return abs(price - level) / level * 100 <= tolerance_pct


def above_level(price: float, level: float | None) -> bool:
    return level is not None and price > level


def below_level(price: float, level: float | None) -> bool:
    return level is not None and price < level


def calc_swing_stop_price(
    side: str,
    swing_high: float | None,
    swing_low: float | None,
    buffer_pct: float,
) -> float | None:
    """Long stop below swing low; short stop above swing high."""
    if side.upper() == "LONG" and swing_low is not None:
        return round(swing_low * (1 - buffer_pct / 100), 2)
    if side.upper() == "SHORT" and swing_high is not None:
        return round(swing_high * (1 + buffer_pct / 100), 2)
    return None


def calc_stop_from_klines(
    klines: list[list],
    side: str,
    *,
    pivot_bars: int,
    lookback: int,
    buffer_pct: float,
) -> float | None:
    highs = [float(k[2]) for k in klines]
    lows = [float(k[3]) for k in klines]
    end_index = len(klines) - 1
    swing_high, swing_low = recent_swing_levels(
        highs, lows, end_index, pivot_bars=pivot_bars, lookback=lookback,
    )
    return calc_swing_stop_price(side, swing_high, swing_low, buffer_pct)
